controllers/auth.js

```diff
@@ -9,6 +9,91 @@ const {
   USER_DOES_NOT_EXIST_ERROR,
 } = require("../constants/errorMessages");
 
+const googleAuthLogin = (req, res, next) => {
+  const { redirectURL, dev } = req.query;
+  if (dev === "true") {
+    return passport.authenticate("google", {
+      scope: ["email"],
+      state: redirectURL,
+    })(req, res, next);
+  } else {
+    return res.boom.unauthorized("User cannot be authenticated");
+  }
+};
+
+async function handleGoogleLogin(req, res, user, authRedirectionUrl) {
+  const rdsUiUrl = new URL(config.get("services.rdsUi.baseUrl"));
+  try {
+    if (!user.emails || user.emails.length === 0) {
+      throw new Error("Email not found in user data");
+    }
+    const userData = {
+      email: user.emails[0].value,
+      created_at: Date.now(),
+      updated_at: null,
+    };
+
+    const userDataFromDB = await users.fetchUser({ email: userData.email });
+
+    if (userDataFromDB.userExists) {
+      if (userDataFromDB.user.roles?.developer) {
+        const errorMessage = encodeURIComponent("Google login is restricted for developer role.");
+        return res.redirect(`${authRedirectionUrl}?error=${errorMessage}`);
+      }
+    }
```

Comment:
Role validation is looking too simplistic. There should be a proper role management system rather than just checking for a boolean flag.

We can do something like this: these improvements would make the code more robust, secure, and maintainable.

`const REQUIRED_AUTH_PROVIDERS = {
  developer: 'github',
  designer: ['google', 'github'],
  product_manager: ['google', 'github']
};

const validateAuthProvider = (userRole, provider) => {
  const allowedProviders = REQUIRED_AUTH_PROVIDERS[userRole];
  return allowedProviders.includes(provider);
};`

Reply:
Currently, we are not assigning specific roles to users, as we have been assuming the roles will be available in the future, based on a discussion with Ankush. 
We plan to implement a proper role collection feature as part of the new-signup process, and once that is in place, we will revisit and improve the role validation logic to reflect the more flexible, secure structure you're proposing.